### DISCUSSION TITLE
FreeBSD deployment (Questions)

### DISCUSSION THREAD

**Discussion by jSML4ThWwBID69YC** [moosefs/moosefs#329] [Q&A]
Hello, 

I'm working to deploy MooseFS on FreeBSD. Are there any deployment guides, or advice available for this? I've read best-practices, but nothing seems FreeBSD specific. 

1: Does ZFS and MooseFS work well together? I'm concerned about MooseFS vs ZFS memory caching conflicting, or wasting resource. 

2: If using ZFS, should MooseFS point to a folder, or a zvol for it's disk? 

3: What is correct value for vfs.fusefs.data_cache_mode? EX: Off, write-through, or write-back.

4: What should mfscachemode be set to on the clients? I recall it was set to DIRECT on earlier version due to bugs. Is this still the case, or is AUTO working on FreeBSD 12.1+?

The current deployment plan is 5x chunk servers with 4x 12TB drives each as a raidz1 setup. MooseFS chunk server configuration will point to a folder for the disk. Any advice, or questions are welcome.

**Comment by chogata** [comment]
Hello,

Ad 1. MooseFS chunk servers rely on OS (local FS) caching, so cache shouldn't be a problem.

Ad 2. Folder. In general, MooseFS works on a/any filesystem, where it can create files and directories with system calls, so no block devices like zvol or any other non-standard volumes.

Ad 3. We tested FreeBSD 12.1 on default settings and they worked well. We did not test any other settings yet.

Ad 4. Auto is ok. MooseFS builds from 3.108 detect whether they're working on old or new version of FUSE api on FreeBSD and adjust accordingly. 

Also, is there a specific reason you want to use raid-z? MooseFS is a fault tolerant system, making RAID redundant in most cases and all our test shows that JBOD setups always work most efficiently.

**Comment by jSML4ThWwBID69YC** [comment]
Hello, 

Thank you for responding. It sounds like everything will just work without tweaking. That's awesome. 

We use ZFS (raidz, raid10) to protect the operating system from disk failure. Using JBOD for the MooseFS disks would still require raid to protect the OS. Using ZFS brings additional features such as snapshots, rollback from upgrades, compression and the adaptive replacement cache (ARC). The ARC should significantly increase read performance allowing us to use larger, but slower disks. 

I can setup a JBOD and ZFS instance to do performance testing and see if there's any significant difference. Before I do, has anybody else used MooseFS on top of ZFS?

**Comment by andrewsali** [comment]
We use moosefs on zfs, we enabled compression to mitigate the inefficiency of Moosefs with small files (the 64kb block size wastes otherwise a lot of space for small files).

**Comment by jSML4ThWwBID69YC** [comment]
Hello @andrewsali 

I am planning to use ZFS compression for the same reason. Most of the workload is small website files.

**Comment by borkd** [comment]
> Hello,
> 
> Thank you for responding. It sounds like everything will just work without tweaking. That's awesome.
> 
> We use ZFS (raidz, raid10) to protect the operating system from disk failure. Using JBOD for the MooseFS disks would still require raid to protect the OS. Using ZFS brings additional features such as snapshots, rollback from upgrades, compression and the adaptive replacement cache (ARC). The ARC should significantly increase read performance allowing us to use larger, but slower disks.
> 
> I can setup a JBOD and ZFS instance to do performance testing and see if there's any significant difference. Before I do, has anybody else used MooseFS on top of ZFS?

For years. Unless you have a very specific application needs and/or scale concerns, I suggest skipping raidz1 and feeding moosefs chunkservers with **single-vdev pools with compression on**. You get to keep the important benefits of ZFS, and let moosefs handle inevitable disk failures.

**Comment by jSML4ThWwBID69YC** [comment]
@borkd 

Thank you for the advice. A single-vdev pool sounds like a good idea. 

For anybody else thinking of using FreeBSD, note the recommendation on https://github.com/moosefs/moosefs/issues/334#issuecomment-579673283. The mfscachemode should be set to DIRECT, even on FreeBSD 12.1.
